File: fastSlow/cycle/detectCycle.js
import {ListNode} from '../../linkedList/ds.js'

// Floyd’s Tortoise and Hare Algorithm
function detect_cycle(head){
    let slow = head;
    let fast = head;

    while(fast && fast.next){
        slow = slow.next;
        fast = fast.next.next;

        if(slow === fast){
            return true;
        }
    }
    return false;
}

const arr = [1, 2, 3, 4, 5]

let head = new ListNode(arr[0])
let curr = head

const nodes = [head]

for (let i = 1; i < arr.length; i++) {
    curr.next = new ListNode(arr[i])
    curr = curr.next
    nodes.push(curr)
}

curr.next = nodes[2]

console.log(detect_cycle(head))

File: fastSlow/cycle/cpp/detectCycle.cpp
#include<iostream>
#include<vector>

using namespace std;

struct Node{
    int data;
    Node* next;

    Node(int value){
        data = value;
        next = nullptr;
    }
};

bool checkCycle(Node* head){
    Node* fast = head;
    Node* slow = head;

    while(fast && fast->next){
        fast = fast -> next -> next;
        slow = slow -> next;
        if(fast && slow){
            return true; 
        }
    }
    return false;
}

int main(){
    vector<int> nums = {0,1,2,3,4};

    Node* head = new Node(nums[0]);
    Node* curr = head;

    for(int i=1;i<nums.size();i++){
        curr -> next = new Node(nums[i]);
        curr = curr -> next;
    }

    curr -> next = head;
    bool ans = checkCycle(head);
    cout<<ans;

    return 0;

}